Cargo.toml

```diff
@@ -11,49 +11,249 @@ readme = "README.md"
+thiserror = { version = "2.0.17", default-features = false }
 
 [dev-dependencies]
-getrandom = { version = "0.2", features = ["custom"] } # workaround to build on no_std targets
+bytes = { version = "1.10.1", default-features = false }
+itertools = { version = "0.14.0", default-features = false }
+rsa_098 = { package = "rsa", version = "0.9.8", features = ["sha2"] }
+signature_220 = { package = "signature", version = "2.2.0" }
+rustls = { version = "0.23.32", default-features = false, features = ["std"] }
+x509-cert = { version = "0.2.5", default-features = false, features = [
+    "builder",
+] }
+rand_core_064 = { package = "rand_core", version = "0.6.4" }
+p256_0132 = { package = "p256", version = "0.13.2" }
 
 [features]
-default = ["std", "tls12", "zeroize"]
+default = ["std", "tls12", "zeroize", "full", "fast", "quic", "ticketer"]
+full = [
+    "aead-full",
+    "sign-full",
+    "verify-full",
+    "kx-full",
+    "hash-full",
+    "format",
+]
+format = ["pem", "pkcs1", "pkcs8", "sec1"]
 logging = ["rustls/logging"]
 tls12 = ["rustls/tls12"]
 
-# Only enable feature in upstream if there is an overall effect e.g. aead/alloc in-place
-# zeroize is another typical that can be turned off
+# RustCrypto is preparing to migrate to core::error::Error
+# and in before most of the use case for std is just std::error::Error
+std = ["alloc", "rustls/std", "ed448-goldilocks?/std", "tinyvec?/std", "thiserror/std"]
+alloc = [
+    "ecdsa?/alloc",
+    "ed448-goldilocks?/alloc",
+    "elliptic-curve?/alloc",
+    "pkcs8?/alloc",
+    "sec1?/alloc",
+    "signature?/alloc",
+]
+zeroize = [
+    "aes-gcm?/zeroize",
+    "aes?/zeroize",
+    "der?/zeroize",
+    "ed25519-dalek?/zeroize",
+    "pkcs1?/zeroize",
+    "sec1?/zeroize",
+    "x25519-dalek?/zeroize",
+]
+subtle = ["digest/subtle", "pkcs8?/subtle", "sec1?/subtle"]
+fast = [
+    "ed25519-dalek?/fast",
+    # "rsa?/u64_digit",
+    "x25519-dalek?/precomputed-tables",
+]
+
+nist = []
+p256 = ["dep:p256", "nist", "p256/pkcs8"]
+p384 = ["dep:p384", "nist", "p384/pkcs8"]
+p521 = ["dep:p521", "nist"]
+ed25519 = ["dep:ed25519-dalek"]
+ed448 = ["dep:ed448-goldilocks"]
+
+ecdsa = ["dep:ecdsa", "verify", "signature", "rand", "der", "elliptic-curve"]
+ecdsa-p256 = ["ecdsa", "p256", "p256/ecdsa"]
+ecdsa-p384 = ["ecdsa", "p384", "p384/ecdsa"]
+ecdsa-p521 = ["ecdsa", "p521", "p521/ecdsa"]
+ecdsa-full = ["ecdsa-p256", "ecdsa-p384", "ecdsa-p521"]
+
+eddsa = ["verify", "signature", "elliptic-curve"]
+eddsa-ed25519 = ["eddsa", "ed25519"]
+eddsa-ed448 = ["eddsa", "ed448"]
+eddsa-full = ["eddsa-ed25519", "eddsa-ed448"]
+
+kx = ["rand", "elliptic-curve"]
+kx-x448 = ["kx", "x448"]
+kx-x25519 = ["kx", "dep:x25519-dalek"]
+kx-nist = ["sec1"]
+kx-p256 = ["kx", "p256", "kx-nist", "p256/ecdh"]
+kx-p384 = ["kx", "p384", "kx-nist", "p384/ecdh"]
+kx-p521 = ["kx", "p521", "kx-nist", "p521/ecdh"]
+kx-full = ["kx-x448", "kx-x25519", "kx-p256", "kx-p384", "kx-p521"]
+
```

Comment:
These features change the behavior of the whole library without bifurcation across the whole build.

In a remote case if there are two things that pull rustcrypto-rustls they are able to add more features transparently to the other thing that pulls only set of limited features.

I'm not sure it's best idea to a) have large feature list b) making things optional this way that should be mandatory and c) where we should enforce mandatory minimums.

It also makes maintenance, reviewability and testing much harder given one needs to test all the possible testable combinations where the feature knob list is tens of items long with each of them having 4-5 or so sub-items.